File: NodeJS/routes/pades-signature.js
const express = require('express');
const path = require('path');
const fs = require('fs');
const uuidv4 = require('uuid/v4');
const { PadesSignatureStarter, StandardSignaturePolicies, SignatureFinisher } = require('pki-express');

const { Util } = require('../util');
const { PadesVisualElements } = require('../pades-visual-elements');

let router = express.Router();
let appRoot = process.cwd();

/**
 * GET /pades-signature
 *
 * This route only renders the signature page.
 */
router.get('/', (req, res, next) => {

	// Verify if the provided userfile exists.
	if (req.query['userfile'] && !fs.existsSync(path.join(appRoot, 'public', 'app-data', req.query['userfile']))) {
		next(new Error('File not found'));
		return;
	}

	res.render('pades-signature', {
		userfile: req.query['userfile']
	});
});

/**
 * POST /pades-signature/start
 *
 * This route starts the signature. In this sample, it will be called
 * programatically after the user press the "Sign File" button (see method
 * readCertificate() on public/javascripts/signature-start-form.js).
 */
router.post('/start', (req, res, next) => {

	let pdfToSign = null;

	// Recover variables from the POST arguments to be use don this step.
	let certThumb = req.body['certThumbField'];
	let certContent = req.body['certContentField'];
	let userfile = req.body['userfileField'];

	// Get an instantiate of the PadesSignatureStarter class, responsible for
	// receiving the signature elements and start the signature process.
	let signatureStarter = new PadesSignatureStarter();

	// Set PKI default options (see util.js).
	Util.setPkiDefaults(signatureStarter);

	// Set signature policy.
	signatureStarter.signaturePolicy = StandardSignaturePolicies.PADES_BASIC_WITH_LTV;

	// Choose the file to be signed. This logic depends if the URL argument
	// "userfile" is set or not.
	if (userfile) {
		// If the URL argument "userfile" is filled, it means the user was
		// redirected here by the file upload.php, this is the "signature with
		// file uploaded by user" case. We'll set the path of the file to be
		// signed, which was saved in the "app-data" folder by upload.php.
		pdfToSign = path.join(appRoot, 'public', 'app-data', userfile);
	} else {
		// If userfile is null, this is the "signature with server file" case.
		// We'll set the path to the sample document.
		pdfToSign = Util.getSampleDocPath();
	}

	// Process all independent IO operations in parallel. This should happen
	// before the method "start" to be called.
	Promise.all([

		// Set PDF to be signed.
		signatureStarter.setPdfToSignFromPath(pdfToSign),

		// Set Base64-encoded certificate's content to signature starter.
		signatureStarter.setCertificateFromBase64(certContent),

		// Set a file reference for the stamp file. Note that this file can be
		// referenced later by "fref://{alias}" at the "url" field on the
		// visual representation (see public/vr.json or
		// getVisualRepresentation() method).
		signatureStarter.addFileReference('stamp', Util.getPdfStampPath()),

		// Set the visual representation. We provided a dictionary that
		// represents the visual representation JSON model.
		signatureStarter.setVisualRepresentation(PadesVisualElements.getVisualRepresentation())

	]).then(() => {

		// Start the signature process. Receive as response the following
		// fields:
		// - toSignHash: The hash to be signed.
		// - digestAlgorithm: The digest algorithm that will inform the Web PKI
		//                    component to compute this signature.
		// - transferFile: A temporary file to be passed to "complete" step.
		return signatureStarter.start();

	}).then(response => {

		// Render the field from start() method as hidden field to be used on
		// the javascript or on the "complete" step.
		res.render('pades-signature-start', {
			toSignHash: response.toSignHash,
			digestAlgorithm: response.digestAlgorithm,
			transferFile: response.transferFile,
			certThumb: certThumb,
			userfile: userfile
		});

	}).catch(err => next(err));
});

/**
 * POST pades-signature/complete
 *
 * This route completes the signature, it will be called programatically after
 * the Web PKI component perform the signature and submit the form (see method
 * sign() on public/javascripts/signature-complete-form.js).
 */
router.post('/complete', (req, res, next) => {

	let outputFile = null;
	let pdfToSign = null;

	// Recover variables from the POST arguments to be used on this step.
	let transferFile = req.body['transferFileField'];
	let signature = req.body['signatureField'];
	let userfile = req.body['userfileField'];

	// Get an instance of the PadesSignatureFinisher class, responsible for
	// completing the signature process.
	let signatureFinisher = new SignatureFinisher();

	// Set PKI default options (see util.js).
	Util.setPkiDefaults(signatureFinisher);

	// Choose the file to be signed. This logic depends if the URL argument
	// "userfile" is set or not.
	if (userfile) {
		// If the URL argument "userfile" is filled, it means the user was
		// redirected here by the file upload.php, this is the "signature with
		// file uploaded by user" case. We'll set the path of the file to be
		// signed, which was saved in the "app-data" folder by upload.php.
		pdfToSign = path.join(appRoot, 'public', 'app-data', userfile);
	} else {
		// If userfile is null, this is the "signature with server file" case.
		// We'll set the path to the sample document.
		pdfToSign = Util.getSampleDocPath();
	}

	// Process all independent IO operations in parallel. This should happen
	// before the method "complete" to be called.
	Promise.all([

		// Set PDF to be signed. It's the same file we used on "start" step.
		signatureFinisher.setFileToSignFromPath(pdfToSign),

		// Set transfer file.
		signatureFinisher.setTransferFileFromPath(transferFile)

	]).then(() => {

		// Set signature.
		signatureFinisher.signature = signature;

		// Generate path for output file and add the signature finisher.
		Util.createAppData(); // Make sure the "app-data" folder exists (util.js).
		outputFile = uuidv4() + '.pdf';
		signatureFinisher.outputFile = path.join(appRoot, 'public', 'app-data', outputFile);

		// Create the signature process.
		return signatureFinisher.complete();

	}).then(() => {

		// Render the result page.
		res.render('pades-signature-complete', {
			outputFile: outputFile
		});

	}).catch(err => next(err));

});

module.exports = router;

File: Python/sample/views/pades_signature.py
"""

This sample performs a PAdES signature in three steps using PKI Express and
Web PKI.

"""
import os
import uuid

from flask import render_template
from flask import request
from flask import current_app
from flask import flash
from flask import redirect
from flask import url_for
from flask import Blueprint
from pkiexpress import standard_signature_policies
from pkiexpress import PadesSignatureStarter
from pkiexpress import SignatureFinisher

from sample.utils import create_app_data
from sample.utils import get_sample_doc_path
from sample.utils import set_pki_defaults
from sample.utils import get_pdf_stamp_path
from sample.utils_pades import get_visual_representation

blueprint = Blueprint('pades_signature', __name__,
                      url_prefix='/pades-signature')


@blueprint.route('/')
@blueprint.route('/<userfile>')
def index(userfile=None):
    """

    This method only renders the signature page.

    """
    # Verify if the provided userfile exists.
    if userfile and not os.path.exists(os.path.join(
            current_app.config['APPDATA_FOLDER'], userfile)):
        return render_template('error.html', msg='File not found')

    return render_template('pades_signature/index.html',
                           userfile=userfile)


@blueprint.route('/start', methods=['POST'])
def start():
    """

    This method starts the signature. In this sample, it will be called
    programatically after the user press the "Sign File" button (see method
    readCertificate() on static/js/signature-start-form.js).

    """
    userfile = None
    try:
        # Recover variables from the POST arguments to be used on this step.
        cert_thumb = request.form['certThumbField']
        cert_content = request.form['certContentField']
        if request.form['userfileField'] != 'None':
            userfile = request.form['userfileField']

        # Get an instance of the PadesSignatureStarter class, responsible for
        # receiving the signature elements and start the signature process.
        signature_starter = PadesSignatureStarter()

        # Set PKI default options (see utils.py).
        set_pki_defaults(signature_starter)

        # Set signature policy.
        signature_starter.signature_policy = \
            standard_signature_policies.PADES_BASIC_WITH_LTV

        # Set PDF to be signed
        if userfile:
            signature_starter.set_pdf_to_sign_from_path(
                os.path.join(current_app.config['APPDATA_FOLDER'], userfile))
        else:
            signature_starter.set_pdf_to_sign_from_path(get_sample_doc_path())

        # Set Base64-encoded certificate's content to signature starter.
        signature_starter.set_certificate_from_base64(cert_content)

        # Set a file reference for the stamp file. Note that this file can be
        # referenced later by "fref://{alias}" at the "url" field on the visual
        # representation (see static/vr.json or get_visual_representation()
        # method).
        signature_starter.add_file_reference('stamp', get_pdf_stamp_path())

        # Set the visual representation. We provided a dictionary that
        # represents the visual representation JSON model.
        signature_starter.set_visual_representation(get_visual_representation())

        # Start the signature process. Receive as response the following fields:
        # - to_sign_hash:     The hash to be signed.
        # - digest_algorithm: The digest algorithm that will inform the Web PKI
        #                     component to compute the signature.
        # - transfer_file:    A temporary file to be passed to "complete" step.
        response = signature_starter.start()

        # Render the field from start() method as hidden field to be used on the
        # javascript or on the "complete" step.
        return render_template('pades_signature/start.html',
                               to_sign_hash=response['toSignHash'],
                               digest_algorithm=response['digestAlgorithm'],
                               transfer_file=response['transferFile'],
                               cert_thumb=cert_thumb,
                               userfile=userfile)

    except Exception as e:
        flash(str(e))
        return redirect(url_for('pades_signature.index', userfile=userfile))


@blueprint.route('/complete', methods=['POST'])
def complete():
    """

    This function completes the signature, it will be called programatically
    after the Web PKI component perform the signature and submit the form (see
    method sign() on static/js/signature-complete-form.js).

    """
    userfile = None
    try:

        # Recover variables form the POST arguments to be used on this step.
        transfer_file = request.form['transferFileField']
        signature = request.form['signatureField']
        if request.form['userfileField'] != 'None':
            userfile = request.form['userfileField']

        # Get an instance of the SignatureFinisher class, responsible for
        # completing the signature process.
        signature_finisher = SignatureFinisher()

        # Set PKI default options (see utils.py).
        set_pki_defaults(signature_finisher)

        # Set PDF to be signed. It's the same file we used on "start" method.
        if userfile:
            signature_finisher.set_file_to_sign_from_path(
                os.path.join(current_app.config['APPDATA_FOLDER'], userfile))
        else:
            signature_finisher.set_file_to_sign_from_path(get_sample_doc_path())

        # Set the transfer file.
        signature_finisher.set_transfer_file_from_path(transfer_file)

        # Set the signature file.
        signature_finisher.signature = signature

        # Generate path for output file and add to the signature finisher.
        create_app_data()  # Guarantees that "app data" folder exists.
        filename = '%s.pdf' % (str(uuid.uuid4()))
        signature_finisher.output_file = \
            os.path.join(current_app.config['APPDATA_FOLDER'], filename)

        # Complete the signature process.
        signer_cert = signature_finisher.complete(get_cert=True)

        return render_template('pades_signature/signature-info.html',
                               signer_cert=signer_cert,
                               filename=filename)

    except Exception as e:
        flash(str(e))
        return redirect(url_for('pades_signature.index', userfile=userfile))
